File: src/content/docs/topics/cs/leetcode-150/graphs/210-course-schedule-ii.py
def find_order(num_courses, prerequisites):
    pass  # TODO: implement

def _run_tests():
    assert find_order(2, [[1, 0]]) == [0, 1]
    result = find_order(4, [[1, 0], [2, 0], [3, 1], [3, 2]])
    assert result.index(0) < result.index(1)
    assert result.index(0) < result.index(2)
    assert result.index(1) < result.index(3)
    assert result.index(2) < result.index(3)
    assert find_order(2, [[1, 0], [0, 1]]) == []
    assert find_order(1, []) == [0]
    result = find_order(3, [])
    assert set(result) == {0, 1, 2}
    assert find_order(3, [[0, 1], [1, 2], [2, 0]]) == []
    # --- large-input timing ---
    import time as _t
    _t0 = _t.perf_counter()
    find_order(200, [[i + 1, i] for i in range(199)])
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f'perf find-order 200 courses DAG chain: {_ms:.1f}ms')
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/graphs/210-course-schedule-ii.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}
final class Solution {
    func findOrder(_ numCourses: Int, _ prerequisites: [[Int]]) -> [Int] {
        // TODO: Implement
        fatalError("TODO: Implement")
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:dc1c058967fd3e5c3a15e30378757b2024dac1fedb307e92bdd3cd5403936a48
    expectEqual(Solution().findOrder(4, [[1, 0], [2, 1], [3, 2]]), [0, 1, 2, 3], "unique-chain")
    expectEqual(Solution().findOrder(2, [[1, 0], [0, 1]]), [], "cycle-has-no-order")
    expectEqual(Solution().findOrder(1, []), [0], "one-course")
    // EXCLUDED_VECTOR course-out-of-range: [2,[[2,0]]] | Every course identifier must be smaller than numCourses.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
